src/org/traccar/reports/Events.java

```diff
@@ -47,16 +67,69 @@ public static String getJson(long userId, Collection<Long> deviceIds, Collection
         return json.build().toString();
     }
 
-    public static String getCsv(long userId, Collection<Long> deviceIds, Collection<Long> groupIds,
-            Collection<String> types, Date from, Date to) throws SQLException {
-        CsvBuilder csv = new CsvBuilder();
-        csv.addHeaderLine(new Event());
+    public static void getExcel(OutputStream outputStream,
+            long userId, Collection<Long> deviceIds, Collection<Long> groupIds,
+            Collection<String> types, Date from, Date to) throws SQLException, IOException {
+        ArrayList<DeviceReport> devicesEvents = new ArrayList<>();
+        ArrayList<String> sheetNames = new ArrayList<>();
+        HashMap<Long, String> geofenceNames = new HashMap<>();
         for (long deviceId: ReportUtils.getDeviceList(deviceIds, groupIds)) {
             Context.getPermissionsManager().checkDevice(userId, deviceId);
+            SortedSet<Event> eventReports = new TreeSet<Event>(new Comparator<Event>() {
+                @Override
+                public int compare(Event e1, Event e2) {
+                    return e1.getServerTime().compareTo(e2.getServerTime());
```

Comment:
There is loop few lines down:
````
for (String type : types) {
    Collection<Event> events = Context.getDataManager().getEvents(deviceId, type, from, to);
````
It requests events for every type separately (except allTypes)
Every result is ordered by DB, but to leave it in natural sorting order for user, whole result have to be sorted too. I do it with `SortedSet`

I can replace this loop by changing SQL request and processing `Collection<String> types` to `('type1', 'type2' ...)`